### ISSUE TITLE
Cloning Dereferenced symbol may cause undefined behaviour.

### ISSUE THREAD

**Issue by Tyleo** [nagisa/rust_libloading#13]
Since adding the Clone functionality I have been experimenting with cloning symbols. I noticed that it is possible to clone a dereferenced symbol. This may cause undefined behavior. In particular this code is broken:

```
let func = {
    let library = Library::new("path_to_lib").unwrap();
    let func = library.get::<fn()>(b"symbol").unwrap();
    let func_ref = &func;
    func_ref.clone()
};
func();
```

This derefs func as `fn()` and calls the clone implementation on `fn()` rather than on the `Symbol` object. Since the `Symbol`'s library has already gone out of scope, the call can fail. I am currently investigating ways to get around this.

The problem seems to be with the `Deref` implementation:

```
impl<T> ::std::ops::Deref for Symbol<T> {
    type Target = T;
    fn deref(&self) -> &T {
        unsafe {
            // Additional reference level for a dereference on `deref` return value.
            mem::transmute(&self.pointer)
        }
    }
}
```

In particular, the transmute drops information about the symbol but allows it to be copied from the raw function pointer.

**Comment by Tyleo** [comment]
Symbol should probably not Deref at all. Rather, it should implement `std::ops::Fn`.

**Comment by nagisa** [comment]
> My current thinking is that Symbol should not implement Deref at all unless it can Deref into an Fn rather than an fn.It should implement std::ops::Fn and also provide a stable call mechanism if possible (std::ops::Fn is currently unstable).

Keep in mind that you can also `dlsym` statics and TLS variables. These must be as usable as the functions.

> This derefs func as fn() and calls the Clone implementation on fn() rather than on the Symbol object. Since the Symbol's Library has already gone out of scope, the call can fail. I am currently investigating ways to get around this.

Ew, that’s nasty.

**Comment by Tyleo** [comment]
I've come to the conclusion that `Symbol`s should not implement `Deref` at all for two reasons:

1. It becomes too easy for a `Deref` coercion to cause confusing behavior. I already described a basic case where a copied `fn()` causes undefined behavior. The general case of this happens for any object with type `T: Copy` which contains a pointer into the loaded library. A user can easily obtain a copy of the symbol which will satisfy incorrect lifetime requirements and outlive the library it is loaded from.
2. The Unsafe guidelines for rust state that, "all functions called from FFI must be marked as unsafe." (https://doc.rust-lang.org/book/unsafe.html)

If `Symbol` must return a reference to its contents, the function call which returns them must be `unsafe`. This rules out `AsRef`, `Deref`, and `Borrow`.

At this point, I think that the best implementation would require different structs for function symbols and data symbols. The function symbols could implement an unsafe method which allows them to be called. They cannot implement `Fn`, even in unstable rust, because it breaks requirement 2 above. The data symbols could implement an unsafe method similar to `Deref` which allows their data to be borrowed. When [negative trait bounds](https://github.com/rust-lang/rfcs/issues/10530) are available, then they would be able to implement `Deref` for `T: !Copy`.

**Comment by nagisa** [comment]
This is blocked on https://github.com/rust-lang/rust/issues/29625.
